doc/Requirements Specification/Parts/Discussion.lyx

```diff
@@ -0,0 +1,283 @@
+
+.
+ Fortunately, it most of the time imposes little loss of generality, as
+ any object oriented software can be described using terms of 
+\begin_inset ERT
+status open
+
+\begin_layout Plain Layout
+
+
+\backslash
+gls{component-based software architecture}
+\end_layout
+
+\end_inset
+
+ (regarding each class as a 
+\begin_inset ERT
+status open
+
+\begin_layout Plain Layout
+
+
+\backslash
+gls{component}
+\end_layout
+
+\end_inset
+
+ in the worst case).
+ Such software will naturally not have the advantages that come with the
+ 
+\begin_inset ERT
+status open
+
+\begin_layout Plain Layout
+
+
+\backslash
+gls{component-based software}
+\end_layout
+
+\end_inset
+
+ approach, but might still be analysed for their performance.
+\begin_inset Note Note
+status open
+
+\begin_layout Plain Layout
+I (Joshua) think this to actually be true, but fear that the SDQ won't agree
+ on this.
+\end_layout
+
+\end_inset
+
+
+\end_layout
+
+\begin_layout Labeling
+\labelwidthstring 00.00.0000
+\begin_inset CommandInset label
+LatexCommand label
+name "/C20/"
+
+\end_inset
+
+ The measured software has a constant, deterministic runtime for a fixed
+ configuration of input parameters, when ignoring influences of the hardware,
+ operating system and error of measurement.
+ This will be the case for most software.
+ The fact the user tries to measure the software when using Beagle implies
+ he expects it to behave in such a manner.
+\end_layout
+
+\begin_layout Section
+Challenges
+\end_layout
+
+\begin_layout Labeling
+\labelwidthstring 00.00.0000
+\begin_inset CommandInset label
+LatexCommand label
+name "/C100/"
+
+\end_inset
+
+ For accurate measurements, the CPU of the test server needs to be controlled.
+ This involves disabling turbo boost, reading the temperature of each core
+ and making sure the CPU is in a real world application thermal state, and
+ possibly further measures.
+\end_layout
+
+\begin_layout Labeling
+\labelwidthstring 00.00.0000
+\begin_inset CommandInset label
+LatexCommand label
+name "/C110/"
+
+\end_inset
+
+ Scalability and transferability need to be ensured.
+\end_layout
+
+\begin_layout Labeling
+\labelwidthstring 00.00.0000
+\begin_inset CommandInset label
+LatexCommand label
+name "/C120/"
+
+\end_inset
+
+ Different hardware platforms vary in different dimensions (CPU frequency,
+ number of CPU cores, size and distribution of CPU caches, speed of RAM,
+ etc.), yet the results have to be representative.
+\end_layout
+
+\begin_layout Labeling
+\labelwidthstring 00.00.0000
+\begin_inset CommandInset label
+LatexCommand label
+name "/C130/"
+
+\end_inset
+
+ Only one component is tested at a time, yet 
+\begin_inset ERT
+status open
+
+\begin_layout Plain Layout
+
+
+\backslash
+glspl{component}
+\end_layout
+
+\end_inset
+
+ need to interact with each other.
+ This means that other 
+\begin_inset ERT
+status open
+
+\begin_layout Plain Layout
+
+
+\backslash
+glspl{component}
+\end_layout
+
+\end_inset
+
+ need to be mocked.
```

Comment:
I don't think so, we could test it like this:
Component A is started at time t0
Component B starts Component A at time t1
Component B returns at time t2
Component A returns at time t3

Then Component B needs time t2-t1 and Component A needs time t3+t1-t2-t0.

On the other hand, this isn't always an option: When A calls B in loop a hundred times, we would measure B a hunderd more times than A.